File: libs/ai-kernel/src/domain/copilot.rs
use uuid::Uuid;

use crate::models::knowledge_base::KnowledgeSearchResult;

#[derive(Debug, Clone)]
pub struct CopilotDraft {
    pub answer: String,
    pub suggested_sql: Option<String>,
    pub pipeline_suggestions: Vec<String>,
    pub ontology_hints: Vec<String>,
}

pub fn assist(
    question: &str,
    dataset_ids: &[Uuid],
    ontology_type_ids: &[Uuid],
    knowledge_hits: &[KnowledgeSearchResult],
    include_sql: bool,
    include_pipeline_plan: bool,
) -> CopilotDraft {
    let lowered = question.to_lowercase();
    let first_dataset = dataset_ids.first().copied();
    let first_ontology_type = ontology_type_ids.first().copied();

    let suggested_sql = if include_sql {
        if let Some(dataset_id) = first_dataset {
            Some(format!(
                "SELECT *\nFROM dataset_{}\nWHERE event_date >= CURRENT_DATE - INTERVAL '30 days'\nORDER BY event_date DESC\nLIMIT 100;",
                dataset_id.simple()
            ))
        } else if lowered.contains("sql") || lowered.contains("query") {
            Some(
				"SELECT *\nFROM your_dataset\nWHERE created_at >= CURRENT_DATE - INTERVAL '7 days';"
					.to_string(),
			)
        } else {
            None
        }
    } else {
        None
    };

    let pipeline_suggestions = if include_pipeline_plan {
        vec![
            "Profile the incoming source and verify schema drift before inference.".to_string(),
            "Materialize embeddings and retrieval chunks as a scheduled upstream step.".to_string(),
            "Add a guardrail validation node before publishing generated outputs.".to_string(),
        ]
    } else {
        Vec::new()
    };

    let mut ontology_hints = Vec::new();
    if let Some(object_type_id) = first_ontology_type {
        ontology_hints.push(format!(
            "Map the response to ontology type {} for downstream actioning.",
            object_type_id.simple()
        ));
    }
    if lowered.contains("ontology") || lowered.contains("object") {
        ontology_hints.push(
            "Prefer stable object identifiers and link types when grounding answers.".to_string(),
        );
    }

    let knowledge_summary = if knowledge_hits.is_empty() {
        "No indexed knowledge passages were required for this answer.".to_string()
    } else {
        format!(
            "Retrieved {} supporting passage(s), starting with '{}'.",
            knowledge_hits.len(),
            knowledge_hits[0].document_title
        )
    };

    CopilotDraft {
        answer: format!(
            "Copilot reviewed the request '{}'. {} Focus the next action on the most recent operational signal and keep the response ready for human verification.",
            truncate(question, 140),
            knowledge_summary
        ),
        suggested_sql,
        pipeline_suggestions,
        ontology_hints,
    }
}

fn truncate(content: &str, limit: usize) -> String {
    let mut chars = content.chars();
    let truncated = chars.by_ref().take(limit).collect::<String>();
    if chars.next().is_some() {
        format!("{truncated}...")
    } else {
        truncated
    }
}

File: openfoundry-go/libs/ai-kernel-go/domain/copilot/copilot.go
// Package copilot hosts the deterministic copilot draft helper —
// pure-logic mirror of libs/ai-kernel/src/domain/copilot.rs. Returns
// the SQL / pipeline / ontology suggestions the chat handler folds
// into a CopilotResponse when the runtime port lands.
package copilot

import (
	"fmt"
	"strings"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/models"
)

type Draft struct {
	Answer              string
	SuggestedSQL        *string
	PipelineSuggestions []string
	OntologyHints       []string
}

func Assist(
	question string,
	datasetIDs []uuid.UUID,
	ontologyTypeIDs []uuid.UUID,
	knowledgeHits []models.KnowledgeSearchResult,
	includeSQL, includePipelinePlan bool,
) Draft {
	lowered := strings.ToLower(question)
	var firstDataset *uuid.UUID
	if len(datasetIDs) > 0 {
		v := datasetIDs[0]
		firstDataset = &v
	}
	var firstOntologyType *uuid.UUID
	if len(ontologyTypeIDs) > 0 {
		v := ontologyTypeIDs[0]
		firstOntologyType = &v
	}

	var suggestedSQL *string
	if includeSQL {
		if firstDataset != nil {
			s := fmt.Sprintf(
				"SELECT *\nFROM dataset_%s\nWHERE event_date >= CURRENT_DATE - INTERVAL '30 days'\nORDER BY event_date DESC\nLIMIT 100;",
				simpleUUID(*firstDataset),
			)
			suggestedSQL = &s
		} else if strings.Contains(lowered, "sql") || strings.Contains(lowered, "query") {
			s := "SELECT *\nFROM your_dataset\nWHERE created_at >= CURRENT_DATE - INTERVAL '7 days';"
			suggestedSQL = &s
		}
	}

	var pipelineSuggestions []string
	if includePipelinePlan {
		pipelineSuggestions = []string{
			"Profile the incoming source and verify schema drift before inference.",
			"Materialize embeddings and retrieval chunks as a scheduled upstream step.",
			"Add a guardrail validation node before publishing generated outputs.",
		}
	}

	ontologyHints := make([]string, 0, 2)
	if firstOntologyType != nil {
		ontologyHints = append(ontologyHints, fmt.Sprintf(
			"Map the response to ontology type %s for downstream actioning.",
			simpleUUID(*firstOntologyType),
		))
	}
	if strings.Contains(lowered, "ontology") || strings.Contains(lowered, "object") {
		ontologyHints = append(ontologyHints,
			"Prefer stable object identifiers and link types when grounding answers.")
	}

	knowledgeSummary := "No indexed knowledge passages were required for this answer."
	if len(knowledgeHits) > 0 {
		knowledgeSummary = fmt.Sprintf(
			"Retrieved %d supporting passage(s), starting with '%s'.",
			len(knowledgeHits), knowledgeHits[0].DocumentTitle,
		)
	}

	return Draft{
		Answer: fmt.Sprintf(
			"Copilot reviewed the request '%s'. %s Focus the next action on the most recent operational signal and keep the response ready for human verification.",
			truncate(question, 140), knowledgeSummary,
		),
		SuggestedSQL:        suggestedSQL,
		PipelineSuggestions: pipelineSuggestions,
		OntologyHints:       ontologyHints,
	}
}

func truncate(content string, limit int) string {
	runes := []rune(content)
	if len(runes) <= limit {
		return content
	}
	return string(runes[:limit]) + "..."
}

// simpleUUID mirrors Rust Uuid::simple() — 32 hex chars without
// hyphens. Lowercase, matching the Rust default formatting.
func simpleUUID(id uuid.UUID) string {
	s := id.String()
	return strings.ReplaceAll(s, "-", "")
}
